lib/redis/distributed/hash_ring.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+require 'zlib'
+
+class Redis
+  class Distributed
+    # Client side consistent hashing
+    class HashRing
+      POINTS_PER_SERVER = 160 # this is the default in libmemcached
+
+      attr_reader :nodes
+
+      # Find the closest index in HashRing with value <= the given value
+      def self.binary_search(ary, value)
```

Comment:
This is subtly diffrent from [Array#bsearch_index](https://ruby-doc.org/core-2.5.3/Array.html#method-i-bsearch_index).